### ISSUE TITLE
Excess property checks disabled when one property matches

### ISSUE THREAD

**Issue by rraval** [Microsoft/TypeScript#30397]
**TypeScript Version:**  3.4.0-dev.20190314

**Search Terms:** excess property

**Code**

```ts
interface Model<Dict extends {}> {
    getDict(): Dict;
}

type GetModelDict<Obj> = Obj extends Model<infer Dict> ? Dict : never;

class User implements Model<{ name: string }> {
    constructor(private readonly name: string) { }

    getDict() {
        return { name: this.name };
    }
}

// f0 is working as intended and correctly triggers excess property checks
function f0(dict: GetModelDict<User>) {
}

f0({foo: "bar"}); // ERROR: `foo` does not exist in type `{name: string}`
f0({name: "Charlie"}); // OKAY
f0({foo: "bob", name: "Charlie"}); // ERROR: `foo` does not exist in type `{name: string}`

// f1 does NOT trigger excess property checks if there is at least one overlapping key
function f1<Obj extends Model<any>, Dict extends GetModelDict<Obj>>(model: Obj, dict: Dict) {
}

const bob = new User("Bob");
f1(bob, {foo: "bar"}); //  ERROR: `foo` does not exist in type `{name: string}`
f1(bob, {name: "Charlie"}); // OKAY
f1(bob, {foo: "bob", name: "Charlie"}); // <----- OKAY??
```

**Expected behavior:**

There should be an excess property error at the line:

```ts
f1(bob, {foo: "bob", name: "Charlie"}); // <----- OKAY??
```

**Actual behavior:**

The compiler allows it.

**Playground Link:** [Playground](https://www.typescriptlang.org/play/#src=interface%20Model%3CDict%20extends%20%7B%7D%3E%20%7B%0A%20%20%20%20getDict()%3A%20Dict%3B%0A%7D%0A%0Atype%20GetModelDict%3CObj%3E%20%3D%20Obj%20extends%20Model%3Cinfer%20Dict%3E%20%3F%20Dict%20%3A%20never%3B%0A%0Aclass%20User%20implements%20Model%3C%7B%20name%3A%20string%20%7D%3E%20%7B%0A%20%20%20%20constructor(private%20readonly%20name%3A%20string)%20%7B%20%7D%0A%0A%20%20%20%20getDict()%20%7B%0A%20%20%20%20%20%20%20%20return%20%7B%20name%3A%20this.name%20%7D%3B%0A%20%20%20%20%7D%0A%7D%0A%0Afunction%20f0(dict%3A%20GetModelDict%3CUser%3E)%20%7B%0A%7D%0A%0Af0(%7Bfoo%3A%20%22bar%22%7D)%3B%0Af0(%7Bname%3A%20%22Charlie%22%7D)%3B%0Af0(%7Bfoo%3A%20%22bob%22%2C%20name%3A%20%22Charlie%22%7D)%3B%0A%0Afunction%20f1%3CObj%20extends%20Model%3Cany%3E%2C%20Dict%20extends%20GetModelDict%3CObj%3E%3E(model%3A%20Obj%2C%20dict%3A%20Dict)%20%7B%0A%7D%0A%0Aconst%20bob%20%3D%20new%20User(%22Bob%22)%3B%0Af1(bob%2C%20%7Bfoo%3A%20%22bar%22%7D)%3B%0Af1(bob%2C%20%7Bname%3A%20%22Charlie%22%7D)%3B%0Af1(bob%2C%20%7Bfoo%3A%20%22bob%22%2C%20name%3A%20%22Charlie%22%7D)%3B%0A)

**Related Issues:**

**Comment by dragomirtitian** [comment]
I belive this is the expected behavior (at least this is behavior I have depended on heavily). If the parameter type is said to `extend` a type then excess property checks are not triggered since the constraint acts as the minimum requirements for that type but do not set any sort of upper limit on what that type could be. 

If you use the type directly as in `f0` that will trigger excess property checks. You can probably use some conditional types if you want to trigger an error on excess properties in this case.

**Comment by rraval** [comment]
@dragomirtitian I cannot speak for whether this behaviour is intentional, but your workaround definitely works (thanks!)

Given the clue that it was the generic `extends` that was disabling the excess property checks, I was able to minimize the code sample further:

```ts
interface User {
    name: string;
}

function f0<Dict extends User>(dict: Dict) {
}

f0({ foo: "bar" }); // ERROR
f0({ name: "Bob" }); // OKAY
f0({ foo: "bar", name: "Bob" }); // NO ERROR

type Strict1<A, B> = Exclude<keyof A, keyof B> extends never ? {} : never;

function f1<Dict extends User>(dict: Dict & Strict1<Dict, User>) {
}

f1({ foo: "bar" }); // ERROR
f1({ name: "Bob" }); // OKAY
f1({ foo: "bar", name: "Bob" }); // ERROR
```

Looking further, I was able to improve the ergonomics of the `Strict` type by removing the need to intersect one of the parameters:

```ts
type Strict2<A, B> = Exclude<keyof A, keyof B> extends never ? A : never;

function f2<Dict extends User>(dict: Strict2<Dict, User>) {
}

f2({ foo: "bar" }); // ERROR
f2({ name: "Bob" }); // OKAY
f2({ foo: "bar", name: "Bob" }); // ERROR
```

This version was good, but I wasn't quite satisfied with the error message:

```ts
f2({ foo: "bar", name: "Bob" }); // Type 'string' is not assignable to type 'never'.
```

Which brings me to the final version which restores the proper error message:

```ts
type Strict3<A, B> = Exclude<keyof A, keyof B> extends never ? A : B;

function f3<Dict extends User>(dict: Strict3<Dict, User>) {
}

f3({ foo: "bar" }); // ERROR
f3({ name: "Bob" }); // OKAY
f3({ foo: "bar", name: "Bob" }); // ERROR
```

The error message is decent:

```ts
f3({ foo: "bar", name: "Bob" });

// Argument of type '{ foo: string; name: string; }' is not assignable to parameter of type 'User'.
//   Object literal may only specify known properties, and 'foo' does not exist in type 'User'.
```

---

I'll leave this open if someone from the TS team wants to chime in about intent, but the workaround is sufficient for my needs.

**Comment by RyanCavanaugh** [comment]
This is the intended behavior.

The excess property checks occur after generic inference, which is why you see an error on the first call - inference failed, so we fell back to the constraint type for the purposes of error reporting, and the object literal's property is in excess of the constraint type.
